server/src/main/java/io/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -0,0 +1,291 @@
+  private final CacheStrategy<T, Object, Query<T>> strategy;
+
+
+  public ResultLevelCachingQueryRunner(
+      QueryRunner baseRunner,
+      QueryToolChest queryToolChest,
+      Query<T> query,
+      ObjectMapper objectMapper,
+      Cache cache,
+      CacheConfig cacheConfig
+  )
+  {
+    this.baseRunner = baseRunner;
+    this.objectMapper = objectMapper;
+    this.cache = cache;
+    this.cacheConfig = cacheConfig;
+    this.query = query;
+    this.strategy = queryToolChest.getCacheStrategy(query);
+    this.populateResultCache = ResultLevelCacheUtil.populateResultLevelCacheOnBrokers(query, strategy, cacheConfig);
+    this.useResultCache = ResultLevelCacheUtil.useResultLevelCacheOnBrokers(query, strategy, cacheConfig);
+  }
+
+  @Override
+  public Sequence<T> run(QueryPlus queryPlus, Map responseContext)
+  {
+    if (useResultCache) {
+
+      final String cacheKeyStr = StringUtils.fromUtf8(strategy.computeCacheKey(query));
+      final byte[] cachedResultSet = fetchResultsFromResultLevelCache(cacheKeyStr);
+      String existingResultSetId = extractEtagFromResults(cachedResultSet);
+
+      existingResultSetId = existingResultSetId == null ? "" : existingResultSetId;
+
+      if (query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH) == null) {
+        query = query.withOverriddenContext(
+            ImmutableMap.of(QueryResource.HEADER_IF_NONE_MATCH, existingResultSetId));
+      }
+      Sequence<T> resultFromClient = baseRunner.run(
+          QueryPlus.wrap(query),
+          responseContext
+      );
+      String newResultSetId = (String) responseContext.get(QueryResource.HEADER_ETAG);
+
+      if (newResultSetId != null && newResultSetId.equals(existingResultSetId)) {
+        log.debug("Return cached result set as there is no change in identifiers for query %s ", query.getId());
+        return deserializeResults(cachedResultSet, strategy, existingResultSetId);
+      } else {
+        @Nullable
+        ResultLevelCachePopulator resultLevelCachePopulator = createResultLevelCachePopulator(
+            cacheKeyStr,
+            newResultSetId
+        );
+
+        if (resultLevelCachePopulator == null) {
+          return resultFromClient;
+        }
+        return Sequences.wrap(Sequences.map(
+            resultFromClient,
+            new Function<T, T>()
+            {
+              @Override
+              public T apply(T input)
+              {
+                cacheResultEntry(resultLevelCachePopulator, input);
+                return input;
+              }
+            }
+        ), new SequenceWrapper()
+        {
+          @Override
+          public void after(boolean isDone, Throwable thrown) throws Exception
+          {
+            Preconditions.checkNotNull(resultLevelCachePopulator, "ResultLevelCachePopulator cannot be null during cache population");
+            if (thrown != null) {
+              log.error("Error while preparing for result level caching for query %s ", query.getId());
```

Comment:
can we add "thrown" to the log argument so that we know what happened .
also what will be the behavior observed by user when this happens ?

Reply:
If an error happens here, cache population would not be performed and the user gets an error message, but the user will still get the query results

Reply:
you mean , error would be reported in the logs and user will get the query results .

Reply:
yes